core/src/main/scala/chisel3/Data.scala

```diff
@@ -552,7 +552,49 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
   /** Whether this Data has the same model ("data type") as that Data.
     * Data subtypes should overload this with checks against their own type.
     */
-  private[chisel3] def typeEquivalent(that: Data): Boolean
+  private[chisel3] final def typeEquivalent(
+    that: Data
+  ): Boolean = findFirstTypeMismatch(that, strictTypes = true, strictWidths = true).isEmpty
+
+  /** Find and report any type mismatches
+    *
+    * @param that Data being compared to this
+    * @param strictTypes Does class of Bundles or Records need to match? Inverse of "structural".
+    * @param strictWidths do widths need to match?
+    * @return None if types are equivalent, Some String reporting the first mismatch if not
+    */
+  private[chisel3] def findFirstTypeMismatch(that: Data, strictTypes: Boolean, strictWidths: Boolean): Option[String] =
```

Comment:
nit: would it be a good idea to provide default values for `strictTypes` and `strictWidth`?

Reply:
I waffled on it a bit, but I actually think "no" because I want any users (and this is an internal API so being pretty doesn't matter) to think about what they are checking. The bug I'm looking to fix (#3200) derives from my telling Jared to use `typeEquivalent` which (as the diff shows) is equivalent to this function with `strictTypes = true` and `strictWidths = true`. If I had thought about that strictness, it would've been obvious that `typeEquivalent` was the wrong function to call.